### PR TITLE
Java frontend: record source locations of read static variables

### PR DESCRIPTION
**Pull request by smowton**
This enables trace-source-location reports not to vary depending on whether static initializer
calls were produced or not (at present the call would result in a statement at the getstatic
location, while the variable read itself would disappear; a 'putstatic' instruction by contrast
always generates an assignment and so is accounted for by a source location).

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e97ed5c11bd481093ee09df0d56e1a6496e0f81a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130069395

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 1]
```diff
@@ -290,6 +290,45 @@ SCENARIO(
...
+        method_symbol.value.visit_pre([&found](const exprt &subexpr) {
+          if(
+            const auto symbol_expr =
+              expr_try_dynamic_cast<symbol_exprt>(subexpr))
+          {
+            if(
+              symbol_expr->source_location().get_java_bytecode_index() == "0" &&
+              symbol_expr->get_identifier() ==
+                "java::ClassReadingStaticField.x")
+              found = true;
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:323]: :pick: multi line condition so please add braces around the body

File: jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp [hunk 2]
```diff
@@ -290,6 +290,45 @@ SCENARIO(
...
+{
+  GIVEN("A class that reads a static field.")
+  {
+    const symbol_tablet symbol_table = load_java_class(
+      "ClassReadingStaticField",
+      "./java_bytecode/java_bytecode_convert_method");
+
+    WHEN("Converting the method that reads said field")
+    {
+      const auto &method_symbol =
```
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_convert_method/convert_method.cpp:306]: :pick: if the symbol isn't found, using lookup_ref makes the output on CI not very clear, better to use lookup and then REQUIRE(method_symbol != nullptr)

**Review by thk123 — APPROVED** [review]
LGTM

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4c5ddc933a0a9887d51cfd6e759eb8e69b2cb40c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130528948

**Pull request merged**